csrc/scheduler/normalization_inner.cpp

```diff
@@ -203,19 +204,29 @@ int64_t getMaxRegisterCountPerThread(
       register_overhead;
 
   // (1) use register calculated from target occupancy
-  int64_t register_per_thread_target = getRegPerThreadGivenThreadsPerSM(
-      target_blocks_per_sm * threads_per_block);
+  int64_t threads_per_sm =
+      std::min(threads_per_block * target_blocks_per_sm, max_threads_per_sm);
+
+  int64_t register_per_thread_target =
+      getRegPerThreadGivenThreadsPerSM(threads_per_sm);
+
   if (register_per_thread_target >= register_per_thread_min) {
-    return register_per_thread_target;
+    return {
+        register_per_thread_target,
+        target_blocks_per_sm * threads_per_block / threads_per_warp};
   }
 
   //(2) can't achieve target occupancy. Estimate occupancy from minimum register
   // each thread should use, then derive register per thread from occupancy.
-  int64_t blocks_per_sm_max =
-      getThreadsPerSMGivenRegPerThread(register_per_thread_min) /
-      threads_per_block;
-  return getRegPerThreadGivenThreadsPerSM(
-      blocks_per_sm_max * threads_per_block);
+  int64_t blocks_per_sm_max = scheduler_utils::safeDiv(
+      getThreadsPerSMGivenRegPerThread(register_per_thread_min),
+      threads_per_block);
+  threads_per_sm =
+      std::min(blocks_per_sm_max * threads_per_block, max_threads_per_sm);
+
+  return {
+      getRegPerThreadGivenThreadsPerSM(threads_per_sm),
+      blocks_per_sm_max * threads_per_block / threads_per_warp};
```

Comment:
we shouldn't be using `blocks_per_sm_max` here any more. If it's capped by `max_threads_per_sm` earlier, our occupancy is also lowered by that as well.

Shouldn't the occupancy just be `threads_per_sm/threads_per_warp`

Reply:
nice catch!

Reply:
:pushpin: ^^^ This change hasn't been applied yet.

Reply:
📌 ^^^ This hasn't been updated yet.

Reply:
updated.